### PR TITLE
Technical users

### PR DESCRIPTION
**Pull request by rmangin-fr**

### CODE REVIEW

File: src/authorization/Authorizations.ts [hunk 1]
```diff
@@ -725,6 +724,10 @@ export default class Authorizations {
     return user.role === UserRole.DEMO;
   }
 
+  public static isTechnical(user: UserToken | User): boolean {
```
Comment (LucasBrazi06) [inline src/authorization/Authorizations.ts:727]: No need to create a method to check a boolean ;-)
Just use user.technical in the IF statements.

File: src/server/rest/v1/service/AuthService.ts [hunk 2]
```diff
@@ -755,6 +755,10 @@ export default class AuthService {
       token = jwt.sign(payload, jwtOptions.secretOrKey, {
         expiresIn: _centralSystemRestConfig.userDemoTokenLifetimeDays * 24 * 3600
       });
+    } else if (Authorizations.isTechnical(user)) {
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/AuthService.ts:758]: use user.technical

File: src/storage/mongodb/UserStorage.ts [hunk 3]
```diff
@@ -455,6 +454,19 @@ export default class UserStorage {
     await Logging.traceEnd(tenant.id, MODULE_NAME, 'saveUserRole', uniqueTimerID);
   }
 
+  public static async saveUserTechnicalFlag(tenant: Tenant, userID: string, flag: boolean): Promise<void> {
```
Comment (LucasBrazi06) [inline src/storage/mongodb/UserStorage.ts:457]: No need a dedicated method.

File: src/server/rest/v1/service/UserService.ts [hunk 4]
```diff
@@ -226,6 +226,8 @@ export default class UserService {
       if (filteredRequest.role) {
         await UserStorage.saveUserRole(req.tenant, user.id, filteredRequest.role);
       }
+      // Save Technical user flag
+      await UserStorage.saveUserTechnicalFlag(req.tenant, user.id, user.technical);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:230]: No need a dedicated method for this flag.
I think that only Admins can see and check this flag, then this should be treated like Admin data (check UserStorage.saveUserAdminData and plateID prop and mimic it)

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/server/rest/v1/service/UserService.ts [hunk 5]
```diff
@@ -232,6 +232,11 @@ export default class UserService {
         adminData.plateID = filteredRequest.plateID || null;
         await UserStorage.saveUserAdminData(req.tenant, user.id, adminData);
       }
+      if (Utils.objectHasProperty(filteredRequest, 'technical')) {
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:235]: saveUserAdminData should be called once for technical and plateID as you did it in createUser :-( !

File: src/server/rest/v1/service/UserService.ts [hunk 6]
```diff
@@ -302,11 +307,14 @@ export default class UserService {
     // Filter
     const filteredRequest = UserValidator.getInstance().validateUserGetByID(req.query);
     UtilsService.assertIdIsProvided(action, filteredRequest.ID, MODULE_NAME, 'handleGetUser', req.user);
+    // Get dynamic auth
+    const authorizationFilter = await AuthorizationService.checkAndGetUserAuthorizations(req.tenant, req.user, { ID: filteredRequest.ID.toString() }, Action.READ);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:311]: No, the method UtilsService.checkAndGetUserAuthorization() should assign the 'projectedFields' to the 'user' right away as AuthorizationService.checkAndGetUserAuthorizations() is already called in UtilsService.checkAndGetUserAuthorization().

File: src/server/rest/v1/service/UserService.ts [hunk 7]
```diff
@@ -302,11 +307,14 @@ export default class UserService {
...
     // Check and Get User
     const user = await UtilsService.checkAndGetUserAuthorization(
       req.tenant, req.user, filteredRequest.ID.toString(), Action.READ, action, null, {
         withImage: true
       }, true, false);
+    user.projectedFields = authorizationFilter.projectFields;
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:317]: To remove

File: src/server/rest/v1/service/UserService.ts (hunk 5 above)
Comment (LucasBrazi06) [reply src/server/rest/v1/service/UserService.ts:235]: user.projectedFields can also be used to replace the IF statements:
`if (Authorizations.isAdmin(req.user) || Authorizations.isSuperAdmin(req.user)) {`

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/server/rest/v1/service/AuthService.ts (hunk 2 above)
Comment (rmangin-fr) [reply src/server/rest/v1/service/AuthService.ts]: done

File: src/authorization/Authorizations.ts (hunk 1 above)
Comment (rmangin-fr) [reply src/authorization/Authorizations.ts]: done

**Review by LucasBrazi06 — APPROVED** [review]

**Pull request merged**